File: libraries/python/ifcx/converters/dxf_writer.py
"""Low-level DXF ASCII writer -- pure Python, no external dependencies.

Provides helpers for emitting group-code/value pairs, sections, tables,
entities, and 3-D points in valid DXF format.
"""

from __future__ import annotations

from contextlib import contextmanager
from typing import Any, Generator


class DxfWriter:
    """Builds a DXF ASCII string incrementally."""

    def __init__(self) -> None:
        self.lines: list[str] = []
        self._handle_counter = 1

    # ------------------------------------------------------------------
    # Primitive writers
    # ------------------------------------------------------------------

    def group(self, code: int, value: Any) -> None:
        """Write a single group-code / value pair.

        The value is formatted according to DXF conventions:
        - Floats use up to 12 significant digits.
        - Booleans become ``0`` or ``1``.
        - Everything else is converted via ``str()``.
        """
        self.lines.append(f"{code:>3}")
        if isinstance(value, float):
            self.lines.append(f"{value:.12g}")
        elif isinstance(value, bool):
            self.lines.append("1" if value else "0")
        elif isinstance(value, int):
            self.lines.append(str(value))
        else:
            self.lines.append(str(value))

    def point(self, x: float, y: float, z: float = 0.0, code_base: int = 10) -> None:
        """Write a 3-D point using consecutive group codes.

        ``code_base`` gives the X code; Y = ``code_base + 10``;
        Z = ``code_base + 20``.
        """
        self.group(code_base, x)
        self.group(code_base + 10, y)
        self.group(code_base + 20, z)

    def handle(self, h: str) -> None:
        """Write a handle (group code 5)."""
        self.group(5, h)

    def next_handle(self) -> str:
        """Allocate and return the next handle as a hex string."""
        h = format(self._handle_counter, "X")
        self._handle_counter += 1
        return h

    def entity(self, entity_type: str) -> None:
        """Write the entity-type marker (group code 0)."""
        self.group(0, entity_type)

    # ------------------------------------------------------------------
    # Structural context managers
    # ------------------------------------------------------------------

    @contextmanager
    def section(self, name: str) -> Generator[None, None, None]:
        """Context manager that emits ``SECTION`` / ``ENDSEC``."""
        self.group(0, "SECTION")
        self.group(2, name)
        yield
        self.group(0, "ENDSEC")

    @contextmanager
    def table(self, name: str, handle: str, entries: int = 0) -> Generator[None, None, None]:
        """Context manager that emits ``TABLE`` / ``ENDTAB``."""
        self.group(0, "TABLE")
        self.group(2, name)
        self.handle(handle)
        self.group(100, "AcDbSymbolTable")
        self.group(70, entries)
        yield
        self.group(0, "ENDTAB")

    # ------------------------------------------------------------------
    # Output
    # ------------------------------------------------------------------

    def to_string(self) -> str:
        """Return the complete DXF content as a string (LF line endings)."""
        return "\n".join(self.lines) + "\n"

File: libraries/csharp/Converters/Dxf/DxfWriter.cs
using System.Globalization;
using System.Text;

namespace Ifcx.Converters.Dxf;

/// <summary>
/// Low-level DXF ASCII writer. Emits group-code/value pairs, sections, tables,
/// entities, and 3D points in valid DXF format.
/// Built from scratch with no external dependencies.
/// </summary>
public sealed class DxfWriter
{
    private readonly StringBuilder _sb = new();
    private int _handleCounter = 1;

    // -----------------------------------------------------------------
    // Primitive writers
    // -----------------------------------------------------------------

    /// <summary>Write a single group-code / value pair.</summary>
    public void Group(int code, object value)
    {
        _sb.Append(code.ToString().PadLeft(3));
        _sb.Append('\n');
        _sb.Append(value switch
        {
            double d => d.ToString("G12", CultureInfo.InvariantCulture),
            float f => f.ToString("G12", CultureInfo.InvariantCulture),
            bool b => b ? "1" : "0",
            int i => i.ToString(),
            _ => value?.ToString() ?? "",
        });
        _sb.Append('\n');
    }

    /// <summary>Write a 3D point using consecutive group codes.</summary>
    public void Point(double x, double y, double z = 0.0, int codeBase = 10)
    {
        Group(codeBase, x);
        Group(codeBase + 10, y);
        Group(codeBase + 20, z);
    }

    /// <summary>Write a handle (group code 5).</summary>
    public void Handle(string h) => Group(5, h);

    /// <summary>Allocate and return the next handle as a hex string.</summary>
    public string NextHandle()
    {
        var h = _handleCounter.ToString("X");
        _handleCounter++;
        return h;
    }

    /// <summary>Write the entity-type marker (group code 0).</summary>
    public void Entity(string entityType) => Group(0, entityType);

    // -----------------------------------------------------------------
    // Structural helpers
    // -----------------------------------------------------------------

    /// <summary>Begin a SECTION.</summary>
    public void BeginSection(string name)
    {
        Group(0, "SECTION");
        Group(2, name);
    }

    /// <summary>End a SECTION.</summary>
    public void EndSection() => Group(0, "ENDSEC");

    /// <summary>Begin a TABLE.</summary>
    public void BeginTable(string name, string handle, int entries = 0)
    {
        Group(0, "TABLE");
        Group(2, name);
        Handle(handle);
        Group(100, "AcDbSymbolTable");
        Group(70, entries);
    }

    /// <summary>End a TABLE.</summary>
    public void EndTable() => Group(0, "ENDTAB");

    // -----------------------------------------------------------------
    // Output
    // -----------------------------------------------------------------

    /// <summary>Return the complete DXF content as a string (LF line endings).</summary>
    public override string ToString() => _sb.ToString();
}
